### DISCUSSION TITLE
Version 1.3.2 released on pypi.org

### DISCUSSION THREAD

**Discussion by Peter-van-Tol** [Peter-van-Tol/LiteX-CNC#36] [Announcements]
After 8 months after the first alpha-releases, there is now finally a stable version available on pypi.org. You can install this version by installing it on your machine:
```shell
pip install litexcnc
```

This version changed a lot in comparison with the alpha-releases:
- fully supports GPIO, PWM/PDM, StepGen, Encoders and can be extended with custom additions (for example a tool-changer). Each module can be configured by using a configuration, or one of the [example configs](https://litex-cnc.readthedocs.io/en/latest/examples/index.html) from the documentation can be used.
- supports 5A-75B and 5A-75E FPGA's (all revisions) with open source toolchain. This toolchain can be automatically installed from the command line with the commands `litexcnc install_litex` and `litexcnc install_toolchain`.
- the FPGA identifies itself, the configuration-file is only required to create the firmware. No additional software is required to run LinuxCNC (in previous versions cJSON was required).
- stable `stepgen` module, which can accept both position and velocity commands. There is no need anymore for external components, usch as `pid` or `pos2vel` (note: it is still possible to use these when required in specific setup, just use the velocity mode in that case).

This release has not been possible without the help of @ozzyrob and OJthe123, for addressing issues, pointing in the right direction and testing the fixes. But also, they showed that their machines work and produce parts. That was great motivation.

**Comment by ghost** [comment]
Congrats mate, that last one needed some hard work on your behalf. 

Must be that time of year, as I was writing this my firewall\router decided it needed some TLC.

**Comment by Peter-van-Tol** [comment]
Version 1.0.0 contained an error due to a merge conflict. A patched version (1.0.1) is now available on pypi.org.

**Comment by Peter-van-Tol** [comment]
Version `v1.0.2.` has been released on pypi.org. In this version the communication error with the FPGA which occurred on some systems has been resolved. Many thanks to OJthe123 for helping tracking down this bug.

**Comment by OJthe123** [comment]
Is there something special for installing on raspberry pi4 with bookworm?

litexcnc install_litex throws errors.

also when I do "pip3 install litexcnc[cli] it doesn't work

```
litexcnc install_litex
Traceback (most recent call last):
  File "/home/cnc/.local/bin/litexcnc", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/cnc/.local/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cnc/.local/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/cnc/.local/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cnc/.local/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cnc/.local/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cnc/.local/lib/python3.11/site-packages/litexcnc/cli/install_litex.py", line 21, in cli
    target = os.path.join(target, 'litex')
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen posixpath>", line 76, in join
TypeError: expected str, bytes or os.PathLike object, not NoneType

```

EDIT:
solved...forgot to install it with '--user'

**Reply by Peter-van-Tol** [reply]
Should also work without the --user directive. Non-critical, but will fix it with the next release.

**Comment by OJthe123** [comment]
I uploaded my setup to the LinuxCNC forum. It might help others to start easy.

https://forum.linuxcnc.org/27-driver-boards/44422-colorcnc-colorlight-5a-75e-5a-75b-as-fpga-controller-board?start=610#281360

**Comment by Peter-van-Tol** [comment]
Just released version 1.0.3 to PyPi. The following bugs have been fixed:
- driver:
  - `encoder`: calculation of average speed has been corrected (#50).
- cli:
  - `install_litex`: Solved bug which prevented installation without the option --user (#47).
  - `install_toolchain`: Auto-detection of both architecture and os. Command line options are only required when installing specific versions.

Also the [documentation](https://litex-cnc.readthedocs.io/en/latest/) have been updated to clarify the usage of the `--user` option on the various commands.

**Comment by Peter-van-Tol** [comment]
Just released version 1.1.0 to PyPi 🍾 🎆 

Added new drivers for SPI communication [48]. For current users this requires a minor modification in their configuration files. The connection entry in the config-file must be altered to include the field `connection_type`. See for further guidelines on the parameters for the connection, including the new SPI connection, the updates help-files.

Further minor changes:
- `cli`:
  - `install_litex`: DEPRECATED, the functionality has been transferred to install_toolchaing.
  - `install_toolchain`: Also installs Litex, and on a RaspberryPi also OpenOCD.
  - `flash_firmware`: Flashes the firmware to the LED-card using OpenOCD. Designed for flashing with a RaspberryPi, however other adapters can also be selected by their name.

**Comment by Peter-van-Tol** [comment]
Just released version 1.2.0 to PyPi 🍾 🎆 

Support for Raspberry Pi 5 added (using spidev only at this moment).

> For this version the configuration file has to be changed. The section `watchdog: {}` has to be added to the configuration in order to compile the firmware. Optionally, one can specify an enable out pin. See for more information the [watchdog documentation ](https://litex-cnc.readthedocs.io/en/latest/modules/watchdog.htm)

- cli:
  - install_toolchain: OpenOCD compiled with support for the libgpiod-driver. Required for support of the Raspberry PI (#64).
  - install_toolchain: Fixed issue with 64-bit OS on RaspberryPi.
  - flash_firmware: Added configuration for libgpiod-driver. Auto-detects whether a Raspberry Pi 5 is used and changes to the new configuration in that case (#64).
- driver:
  - watchdog: Watchdog requires configuration. Optional an enable out pin can be set for the Watchdog (#65).
  - HUB75HAT: Corrected pinout (#68)
  - gpio: Fixed issue with configurations with either all inputs or all outputs (#70)

**Comment by Peter-van-Tol** [comment]
**Just released version 1.2.2 to PyPi 🍾 🎆**

Bugfix version for stepgen module, which was affected by an error in the encoder module.

Due to an upgrade in the detection / comparison of the version of the firmware this version will show a warning message if older firmware is used (NOTE: the communication has not changed, thus re-compilation is not required). In the next minor release (i.e. version 1.3) this change will correctly enforce recompilation of the firmware. A change in minor version (i.e. 1.2 -> 1.3) indicates a modification in communication protocol and thus requires an update of the firmware.

- driver:
  - `encoder`: counts and position of encoder can be reset (#74)
  - `encoder`: incorrect required read buffer reported has been corrected. This affected the module stepgen (#79)
  - `watchdog`: the watchdog is reset when the card is reset. Prevents the has_bitten message when LinuxCNC is restarted without power-cycling the card. (#80)
  - Fixed showing the correct version of LitexCNC in HAL.
  - Temporary fixed comparison between version of FPGA and driver. When the minor version changes in an upcoming release, the driver will now correctly enforce a recompilation of the firmware.
- cli:
  - `install_toolchain`: Fixed issue with detection of OS on desktop. Also `git` and `openocd` are installed on x64-architecture systems (#78).

**Comment by Peter-van-Tol** [comment]
**Just released version 1.3.0 to PyPi 🍾 🎆**

In 2024 a lot of improvements have been made with the help of input by you! Thank you for helping in making this a better project. All changes have been compiled in the version 1.3.0 release.

> For this version the configuration file has to be changed. The section ``watchdog: {}`` has been expanded with extra functions. See: `watchdog documentation <https://litex-cnc.readthedocs.io/en/latest/modules/watchdog.html>`_ 

* ``firmware``:

  * ``watchdog``: completly new watchdog. Functions can be added to the ``watchdog``, such as a heartbeat or an enable signal. The enable signal can for example be used to disable the stepper drivers when the watchdog has detected an error. Also E-stop can be directly be wired into the ``watchdog``, making them an integral part of the E-stop chain.
  * ``stepgen``: the ``max_frequency`` of the stepgen can be set. The default value is 400 kHz (equal to prior v ersions). With the ``max_frequency`` and the clock frequency of the FPGA the optimal shift is calculatedensuring the maximum resolution, whilst assuring the requested maximum frequency. 
  * ``pwm``: Supports two additional modes: ``PWM/direction`` and ``UP/DOWN``.
  * ``encoder``: Supports counter mode.
  * ``gpio``: Fixed bug in building firmware when all ``gpio`` are either input or output.
  * the LED and button on the 5A-75B and 5A-75E can be used in the firmware.
  * toolchain: fixed bug in detecting OS.
  * toolchain: bumped version of Litex to 2024.12.
  * toolchain: bumped version of OSS-cad-suite to 20241231.

* ``driver``:

  * ``watchdog``:  completly new watchdog. The driver watchdog is implemented as a E-stop latch, so it can be used within an E-stop chain in the HAL. When the watchdog has bitten, it has to be actively reset by resetting the E-stop latch to resume command of the FPGA. Examples on how to tie the ``watchdog`` into HAL are provided in the documentation.
  * ``stepgen``: Added ``index-enable`` and ``index-pulse``. When the ``index-enable`` is HIGH and a pulse is received on ``index-pulse`` the counter of the ``stepgen`` is reset to 0.
  * ``pwm``: the signal of PWM can be inverted.
  * all pin names are now consistently using ``-`` as seperator (i.e. ``index-enable`` instead of ``index_enable``)

**Reply by Peter-van-Tol** [reply]
**Just released version 1.3.1 to PyPi** 🍾 🎆

The script `litexcnc` was not installed on all systems due to a new version of the build-system. This issue has been fixed in this version. No functional changes with the prior version.

**Reply by Peter-van-Tol** [reply]
**Just released version 1.3.2 to PyPi** 🍾 🎆
An error in ``stepgen`` has been resolved in this version. Upgrading from version 1.3.0 or 1.3.1 to this version requires re-compilation of the firmware and re-installation of the drivers.

* `firmware`:
  * `stepgen`: config on the stepgen was not correctly stored on the FPGA, hence the defined stepgens were not detected by the driver and with more then 3 stepgens errors were generated if encoders were also defined in the configuration.
* `driver`:
  * ``stepgen``: when using 3, 7 , 11, and so on stepgens, the ``encoder`` module would fail.
